main.js

```diff
@@ -216,6 +216,10 @@ function createWindow() {
   mainWindow = new BrowserWindow(windowOptions);
 
   function captureAndSaveWindowStats() {
+    if (!mainWindow) {
+      return;
+    }
```

Comment:
If we don’t expect this to happen often, would it make sense to add a log entry?

Reply:
Looking at the code more closely, how does `mainWindow` become `null` after it’s been set? Maybe this could point to a deeper problem.

Is it because of this debounce that fires in a delayed manner? https://github.com/signalapp/Signal-Desktop/blob/a0da73ca8d60d59b4687e60e44f85bca4ef58b07/main.js#L242 If so, could we detach the event listeners when `mainWindow` is set to `null`?

Reply:
I've run into it when Ctrl-C'ing the app. And the user report came from timing issues around coming back from sleep.

Looking at the code, the first possibility is if this code runs before `mainWindow` is initially created. The second is between window close and the app going away. Neither are high-value cases. I think we can leave it as is.

```
  // Emitted when the window is closed.
  mainWindow.on('closed', () => {
    // Dereference the window object, usually you would store windows
    // in an array if your app supports multi windows, this is the time
    // when you should delete the corresponding element.
    mainWindow = null;
  });
```

Reply:
Thanks for sharing that snippet, I must have missed it. At some point, it might make sense to remove all event handlers that rely on `mainWindow` to be set so we’re a bit more explicit about the lifecycle of the app.

Reply:
Re: this code might run first. Is that even possible given that `captureAndSaveWindowStats` is only added as event handler _after_ `mainWindow` has been initialized?

Reply:
Yeah, I think you're right; it's only the second case I listed above.